XMonad/Actions/EasyMotion.hs

```diff
@@ -0,0 +1,332 @@
+                                   rh = max (fi h) (fi th)
+
+-- | Use to create overlay windows the minimum size to contain their key chord
+textSize :: Position -> Rectangle -> Rectangle
+textSize th r = Rectangle { rect_width  = fi th
+                          , rect_height = fi th
+                          , rect_x      = rect_x r + (fi (rect_width r) - fi th) `div` 2
+                          , rect_y      = rect_y r + (fi (rect_height r) - fi th) `div` 2 }
+
+-- | Use to create overlay windows the full width of the window they select, the minimum height to
+--   contain their chord, and a proportion of the distance from the top of the window they select
+bar :: RealFrac f => f -> Position -> Rectangle -> Rectangle
+bar f th r = Rectangle { rect_width  = rect_width r
+                       , rect_height = fi th
+                       , rect_x      = rect_x r
+                       , rect_y      = rect_y r + round (f' * (fi (rect_height r) - fi th)) }
+                         -- clamp f in [0,1] as other values will appear to lock up xmonad
+                         -- as the overlay will be displayed off-screen
+                         where f' = min 0.0 $ max f 1.0
+
+-- | Handles overlay display and window selection. Called after config has been sanitised.
+handleSelectWindow :: EasyMotionConfig -> X (Maybe Window)
+handleSelectWindow c = do
+  f <- initXMF $ font c
+  th <- textExtentsXMF f (concatMap keysymToString (concat $ sKeys c)) >>= \(asc, dsc) -> return $ asc + dsc + 2
+  XConf { theRoot = rw, display = dpy } <- ask
+  XState { mapped = mappedWins, windowset = ws } <- get
+  let currentW = W.stack . W.workspace . W.current $ ws
+      -- Bucket windows by screen, unless the user has provided only a single list of keys
+      winBuckets :: X [[Overlay]]
+      winBuckets = sequence $ fmap (sequence . fmap buildOverlay) $
+        case sKeys c of
+          [_] -> [toList mappedWins]
+          _ -> map (L.filter (`elem` (toList mappedWins)) . W.integrate' . W.stack . W.workspace) sortedScreens
+            where
+              sortedScreens = sortOn ((rect_x &&& rect_y) . screenRect . W.screenDetail) (W.current ws : W.visible ws)
+      -- Zip window buckets with selection keys, then sort them within screens
+      sortedOverlays :: X [[Overlay]]
+      sortedOverlays = fmap (zipWith (appendChords (maxChordLen c)) (sKeys c) . fmap (sortOn ((wa_x &&& wa_y) . attrs))) winBuckets
+      displayF = displayOverlay f (bgCol c) (borderCol c) (txtCol c) (borderPx c)
+      buildOverlay :: Window -> X Overlay
+      buildOverlay w = do
+        wAttrs <- io $ getWindowAttributes dpy w
+        let r = overlayF c th $ makeRect wAttrs
+        o <- createNewWindow r Nothing "" True
+        return Overlay { rect=r, overlay=o, win=w, attrs=wAttrs, chord=[] }
+  overlays <- fmap concat sortedOverlays
+  status <- io $ grabKeyboard dpy rw True grabModeAsync grabModeAsync currentTime
+  if (status == grabSuccess)
```

Comment:
Some of these would perhaps be more readable in a `where` clause (and with space separating individual functions)

Reply:
Sheesh, I've made.. some.. changes.. I don't have a good feel for Haskell aesthetics yet. E.g. `appendChords` is off by itself, it's a little larger than the others I've put in a `where` clause, but not by much. I'll resolve this particular item for now because the code has changed quite a bit since the above comment. More feedback in this vein very welcome, though.

Reply:
I think `appendChords` being a top-level function is not that bad; don't really want `handleSelectWindow` to become even bigger!  Btw, because I was just looking at this, you should be able to write 

``` Haskell
zipWith Overlay overlayWins chords
```

instead of having to explicitly deal with record syntax there.

Regarding the style I'll jot down a few things in the second review; particularly about those 150+ column lines ;)